src/qibolab/pulses/pulse.py

```diff
@@ -152,15 +117,31 @@ def __hash__(self):
             )
         )
 
-    def is_equal_ignoring_start(self, item) -> bool:
-        """Check if two pulses are equal ignoring start time."""
-        return (
-            self.duration == item.duration
-            and self.amplitude == item.amplitude
-            and self.frequency == item.frequency
-            and self.relative_phase == item.relative_phase
-            and self.shape == item.shape
-            and self.channel == item.channel
-            and self.type == item.type
-            and self.qubit == item.qubit
-        )
+
+@dataclass
+class Delay:
+    """A wait instruction during which we are not sending any pulses to the
+    QPU."""
+
+    duration: int
+    """Delay duration in ns."""
+    channel: str
+    """Channel on which the delay should be implemented."""
+    type: PulseType = PulseType.DELAY
+    """Type fixed to ``DELAY`` to comply with ``Pulse`` interface."""
+
+
+@dataclass
+class VirtualZ:
+    """Implementation of Z-rotations using virtual phase."""
+
+    duration = 0
+    """Duration of the virtual gate should always be zero."""
+
```

Comment:
Then why do you want to have this attribute at all?

You could document this behavior in the docstring, and drivers should act consequently, isn't it?

Unless we want to define a common interface, beyond `Pulse` itself.

Reply:
> Unless we want to define a common interface, beyond `Pulse` itself.

Indeed, the problem is in this sentence. In this case I added it because `sequence.duration` calls `pulse.duration` and it wouldn't be usable if the `sequence` contains `VirtualZ` pulses and they do not have `duration` attribute. 

However, the bigger problem comes from the different types that one can `append` to a `PulseSequence`. Currently here we have:
```py
class Pulse:
    ...

class Delay:
    ...

class VirtualZ:
    ...
```
and `PulseSequence: list[Pulse | Delay | VirtualZ]`. Then we should somehow make sure that all methods under `PulseSequence` comply with all three types. Personally, I would prefer something more organized:
```py
class SequenceElement:  # (maybe we need a better name)
    ...

class Pulse(SequenceElement):
    ...

class Delay(SequenceElement):
    ...

class VirtualZ(SequenceElement):
    ...
```
and `PulseSequence: list[SequenceElement]` with `SequenceElement` defining everything that `PulseSequence` requires. We could even promote `Pulse` to `SequenceElement` and make `Delay` and `VirtualZ` inherit it, the price would be that they would have a few methods/attributes that they would not need.

@alecandido @hay-k let me know what you think.

Reply:
To be fair, that's the reason why I wanted all of them to be just `Pulse`s, and split the `Delay` and `VirtualZ` (at the time it was only `Delay`), just by using the shape attribute. Though it's definitely not a shape...